### ISSUE TITLE
array-callback-return should check forEach

### ISSUE THREAD

**Issue by SheetJSDev** [eslint/eslint#12551]
**What rule do you want to change?**
array-callback-return

**Does this change cause the rule to produce more or fewer warnings?**
more

**How will the change be implemented? (New option, new default behavior, etc.)?**
new default behavior

**Please provide some example code that this change will affect:**

```js
[
    "abc",
    "def"
].forEach((word) => {

    "use strict";

    return word;

});
```

**What does the rule currently do for this code?**
Pass

**What will the rule do after it's changed?**
Warn that the forEach callback potentially returns a value 

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by g-plane** [comment]
Thanks for this issue.

Can you provide more information or context to tell us the intention of this change?

**Comment by SheetJSDev** [comment]
[From the documentation for the rule](https://eslint.org/docs/rules/array-callback-return)

> Array has several methods for filtering, mapping, and folding. If we forget to write return statement in a callback of those, it’s probably a mistake.  If you don’t want to use a return or don’t need the returned results, consider using .forEach instead.

It is probably a mistake for callbacks of the forEach method to return values (since forEach does not use the result).  If you want to use a return or need the returned result, consider using `.map` instead.

forEach / map confusion is not uncommon.   The rule checks one direction (map used when forEach was more appropriate) but not the other direction (forEach used when map is more appropriate)

**Comment by g-plane** [comment]
However, this rule warns if you forget to return something when using those methods which accept callback, while you want to propose about `forEach`. It shouldn't be a behavior of this rule.

**Comment by SheetJSDev** [comment]
There are two questions you can ask:

1) should ESLint warn if a `forEach` callback returns a value?

since `array-callback-return` currently checks callbacks for other array functions like `map`, it would be consistent for ESLint to check the callbacks for `forEach`.  Array#forEach is a commonly used method. 

2) should the `forEach` check be part of `array-callback-return` or a new rule?

The name `array-callback-return` and the quoted language from the documentation suggests the general intent of the rule is to detect common mistakes in return values from callbacks of Array methods.  As `forEach` is an Array method, it makes the most sense to change the rule rather than introducing a new rule.

**Comment by ljharb** [comment]
The name of the rule suggests it should handle both cases - it’s just only doing half its job right now. An option to check the inverse case on forEach seems ideal.

**Comment by kaicataldo** [comment]
I’ll champion this. While this would be a change of scope, I agree that this is in the spirit of the rule and is useful. I don’t think we should create another rule for this.

That being said, I think this has to go behind an option that is turned off by default and that we should make it the default behavior in a future major release.

**Comment by platinumazure** [comment]
This is now accepted.

**Comment by mdjermanovic** [comment]
PR #12646
